data-prepper-api/src/main/java/org/opensearch/dataprepper/model/source/coordinator/SourceCoordinator.java

```diff
@@ -29,44 +29,56 @@
      *
      * @return {@link SourcePartition} with the details about how to process this partition. Will repeatedly return the partition until
-     * {@link SourceCoordinator#completePartition(PartitionIdentifier)}
-     * or {@link SourceCoordinator#closePartition(PartitionIdentifier, Duration)} are called by the source,
+     * {@link SourceCoordinator#completePartition(String)}
+     * or {@link SourceCoordinator#closePartition(String, Duration)} are called by the source,
      * or until the partition ownership times out.
      * @since 2.2
      */
     Optional<SourcePartition<T>> getNextPartition();
 
     /**
      * Should be called by the source when it has fully processed a given partition
-     * @param partitionIdentifier - The partition that identifies the partition of work that was fully processed
+     *
+     * @param partitionKey - The partition key that uniquely identifies the partition of work that was fully processed
+     * @return Whether the partition was marked as completed successfully or not. If this returns false, the operation is safe to retry.
      * @since 2.2
      */
-    void completePartition(final PartitionIdentifier partitionIdentifier);
+    boolean completePartition(final String partitionKey);
 
     /**
      * Should be called by the source when it has processed all that it can up to this point in time for a given partition,
      * but the partition is expected to need work done on it in the future. While not required, it is necessary in most cases to call saveStateForPartition before closing it.
      * Otherwise, when the partition reopens, the state will be outdated and duplicate parts of the partition will be processed.
-     * @param partitionIdentifier - The partition that identifies the partition of work
-     * @param reopenAfter - The duration from the current time to wait before this partition should be processed further at a later date
+     *
+     * @param partitionKey - The partition key that uniquely identifies the partition of work
+     * @param reopenAfter  - The duration from the current time to wait before this partition should be processed further at a later date
+     * @return Whether the partition was closed successfully or not. If this returns false, the operation is safe to retry.
+     * @throws org.opensearch.dataprepper.model.source.coordinator.exceptions.PartitionNotFoundException if the partition key could not be found in the distributed store
+     * @throws org.opensearch.dataprepper.model.source.coordinator.exceptions.PartitionNotOwnedException if the partition is not owned by this instance of SourceCoordinator
      * @since 2.2
      */
-    void closePartition(final PartitionIdentifier partitionIdentifier, final Duration reopenAfter);
+    boolean closePartition(final String partitionKey, final Duration reopenAfter);
```

Comment:
Could you elaborate on the use case for closing a partition before it is fully processed?

Reply:
Yes. So closing a partition should only happen to a point where the whole partition has been processed that can be up to this point in time. So in the case of an s3 object, we probably want to just complete the partition as we processed the whole thing.

However, this gives sources the option to not mark the partition as completed, and say that more data will be included in this partition in the future, and at some point we should reopen it to process the new data. 

An example of this could be an OpenSearch index that we process all the documents for that exist at a certain point in time, and then at some point in the future, we want to process this index again to pick up new documents that have been indexed after the intitial processing of the index.